### PR TITLE
TTL business validator

### PR DESCRIPTION
**Pull request by AnPogrebnyak**
Rejects or warning messages if them exceeded configured TTL parameter.
In case of TTL parameters are not configured, rule is ignored.
In case of configured "action" value is NONE, rule is ignored.
In case of configured "action" value is missing, rule is ignored.
In case of configured "action" value is REJECT, rule behavior is rejecting message.
In case of configured "action" value is WARNING, rule behavior is log warning.

All cases are covered by the tests.

### CODE REVIEW

File: src/main/java/com/icthh/xm/tmf/ms/communication/rules/ttl/TTLRule.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+    @Override
+    public RuleResponse validate(CommunicationMessage message) {
+        RuleResponse ruleResponse = new RuleResponse();
+        ruleResponse.setRuleType(TTL_RULE);
+        ruleResponse.setResponseCode(TTL_EXCEED_MESSAGE_CODE);
+        ruleResponse.setSuccess(true);
+        TTLRuleConfig config = configService.getTtlRuleConfig();
+        if (config != null && config.isActive() && message.getCharacteristic() != null) {
+            config.getTTL().ifPresent((ttl) ->
```
Comment (amedvedchuk) [inline src/main/java/com/icthh/xm/tmf/ms/communication/rules/ttl/TTLRule.java]: too long lambda, suggest to extract to method

File: src/main/java/com/icthh/xm/tmf/ms/communication/rules/ttl/TTLRule.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+                message.getCharacteristic().stream()
+                    .filter((characteristic) -> MESSAGE_RECEIVED_BY_CHANNEL_TIMESTAMP.equals(characteristic.getName()))
+                    .peek((characteristic) -> log.debug("The following bornTime is found: {}", characteristic.getValue()))
+                    .findAny()
+                    .filter((characteristic) -> characteristic.getValue() != null)
+                    .map((characteristic) -> new Date(Long.valueOf(characteristic.getValue())).toInstant())
+                    .filter((bornTime) -> Instant.now().minus(ttl).isAfter(bornTime))
+                    .ifPresent((bornTime) -> {
```
Comment (amedvedchuk) [inline src/main/java/com/icthh/xm/tmf/ms/communication/rules/ttl/TTLRule.java]: too long lambda, suggest to extract to method

File: src/main/java/com/icthh/xm/tmf/ms/communication/rules/ttl/TTLRule.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    public RuleResponse validate(CommunicationMessage message) {
+        RuleResponse ruleResponse = new RuleResponse();
+        ruleResponse.setRuleType(TTL_RULE);
+        ruleResponse.setResponseCode(TTL_EXCEED_MESSAGE_CODE);
+        ruleResponse.setSuccess(true);
+        TTLRuleConfig config = configService.getTtlRuleConfig();
+        if (config != null && config.isActive() && message.getCharacteristic() != null) {
+            config.getTTL().ifPresent((ttl) ->
+                message.getCharacteristic().stream()
+                    .filter((characteristic) -> MESSAGE_RECEIVED_BY_CHANNEL_TIMESTAMP.equals(characteristic.getName()))
```
Comment (amedvedchuk) [inline src/main/java/com/icthh/xm/tmf/ms/communication/rules/ttl/TTLRule.java]: seems you can add one `map` operation after filter as:
```
.map(CommunicationRequestCharacteristic::getValue)
```
it should eliminate further .getValue() calls.

File: src/test/java/com/icthh/xm/tmf/ms/communication/config/TenantConfigMockConfiguration.java [hunk 4]
```diff
@@ -42,4 +43,9 @@ public TenantConfigRepository tenantConfigRepository() {
     public BusinessTimeConfigService businessTimeConfigService() {
         return mock(BusinessTimeConfigService.class);
     }
+
+    @Bean("TTLRuleConfigService")
```
Comment (amedvedchuk) [inline src/test/java/com/icthh/xm/tmf/ms/communication/config/TenantConfigMockConfiguration.java]: why do we need bean name definition here?

File: src/main/java/com/icthh/xm/tmf/ms/communication/config/KafkaChannelFactory.java [hunk 5]
```diff
@@ -151,4 +152,27 @@ private CommunicationMessage mapToCommunicationMessage(String eventBody) {
         return objectMapper.readValue(eventBody, CommunicationMessage.class);
     }
 
+    /**
+     * Since Kafka headers are not accessible from the business rules,
+     * move Kafka received timestamp to the communication message characteristics
+     */
+    private void addReceivedByChannelCharacteristic(CommunicationMessage communicationMessage, Message<?> kafkaMessage){
+        // cannot be null, but anyway
+        if (kafkaMessage.getHeaders() != null){
```
Comment (amedvedchuk) [inline src/main/java/com/icthh/xm/tmf/ms/communication/config/KafkaChannelFactory.java]: please consider if it worth to use functional approach with Optional like this:
```
Optional.ofNullable(kafkaMessage)
   .map(Message::getHeaders)
   .map(headers -> headers.get(KafkaHeaders.RECEIVED_TIMESTAMP))
   .map(Long::cast)
   .ifPresent(time -> 
                    communicationMessage.addCharacteristicItem(
                    new CommunicationRequestCharacteristic()
                        // Rename it to unlink name from source channel
                        .name(MESSAGE_RECEIVED_BY_CHANNEL_TIMESTAMP)
                        .value(String.valueOf(kafkaReceivedTimestamp))
                );
)
```

File: src/test/java/com/icthh/xm/tmf/ms/communication/config/TenantConfigMockConfiguration.java (hunk 4 above)
Comment (AnPogrebnyak) [reply src/test/java/com/icthh/xm/tmf/ms/communication/config/TenantConfigMockConfiguration.java]: yes, spring manages this bean as "TTLRuleConfigService" and cannot find the factory method without annotation. Just try to delete it and run tests

File: src/main/java/com/icthh/xm/tmf/ms/communication/rules/ttl/TTLRule.java (hunk 2 above)
Comment (AnPogrebnyak) [reply src/main/java/com/icthh/xm/tmf/ms/communication/rules/ttl/TTLRule.java]: Done

File: src/main/java/com/icthh/xm/tmf/ms/communication/rules/ttl/TTLRule.java (hunk 3 above)
Comment (AnPogrebnyak) [reply src/main/java/com/icthh/xm/tmf/ms/communication/rules/ttl/TTLRule.java]: Done

File: src/main/java/com/icthh/xm/tmf/ms/communication/config/KafkaChannelFactory.java (hunk 5 above)
Comment (AnPogrebnyak) [reply src/main/java/com/icthh/xm/tmf/ms/communication/config/KafkaChannelFactory.java]: Done

File: src/main/java/com/icthh/xm/tmf/ms/communication/rules/ttl/TTLRule.java (hunk 1 above)
Comment (AnPogrebnyak) [reply src/main/java/com/icthh/xm/tmf/ms/communication/rules/ttl/TTLRule.java]: Done

**Pull request merged**